pkg/cloud-controller-manager/loadbalancer.go

```diff
@@ -0,0 +1,371 @@
+		lb.Status.Address = service.Status.LoadBalancer.Ingress[0].IP
+	}
+
+	return lb
+}
+
+func (l *LoadBalancerManager) updateServiceLoadBalancerIP(lbName string, service *v1.Service) error {
+	object, ip, err := waitForIP(func() (runtime.Object, string, error) {
+		lb, err := l.lbClient.Get(l.namespace, lbName, metav1.GetOptions{})
+		if err != nil || lb.Status.AllocatedAddress.IP == "" {
+			return nil, "", fmt.Errorf("could not get allocated IP address")
+		}
+		return lb, lb.Status.AllocatedAddress.IP, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	lb := object.(*lbv1.LoadBalancer)
+	if service.Annotations != nil && service.Annotations[KeyKubevipLoadBalancerIP] == ip && lb.Status.Address == ip {
+		return nil
+	}
+
+	serviceCopy := service.DeepCopy()
+	if serviceCopy.Annotations == nil {
+		serviceCopy.Annotations = make(map[string]string)
+	}
+	serviceCopy.Annotations[KeyKubevipLoadBalancerIP] = ip
+	if _, err := l.localSvcClient.Update(serviceCopy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LoadBalancerManager) updateSecondaryServiceLoadBalancerIP(ip string, service *v1.Service) error {
+	if service.Annotations != nil && service.Annotations[KeyKubevipLoadBalancerIP] == ip && service.Annotations[KeyIPAM] == "" {
+		return nil
+	}
+
+	serviceCopy := service.DeepCopy()
+	if serviceCopy.Annotations == nil {
+		serviceCopy.Annotations = make(map[string]string)
+	}
+	// update the annotations and kube-vip will update the service status load balancer
+	serviceCopy.Annotations[KeyKubevipLoadBalancerIP] = ip
+	delete(serviceCopy.Annotations, KeyIPAM)
+	if _, err := l.localSvcClient.Update(serviceCopy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LoadBalancerManager) deleteLoadBalancer(clusterName string, service *v1.Service) error {
+	// check if there are other services using the same load balancer
+	if err := l.checkSecondaryServicesBeforeDeleted(service); err != nil {
+		return fmt.Errorf("could not delete load balancer for service %s/%s: %w", service.Namespace, service.Name, err)
+	}
+
+	name := loadBalancerName(clusterName, service.Namespace, service.Name, string(service.UID))
+	_, err := l.lbClient.Get(l.namespace, name, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if err == nil {
+		return l.lbClient.Delete(l.namespace, name, &metav1.DeleteOptions{})
+	}
+
+	return nil
+}
+
+func waitForIP(callback func() (runtime.Object, string, error)) (runtime.Object, string, error) {
+	for i := 0; i < retryTimes; i++ {
+		object, ip, err := callback()
+		if err == nil {
+			return object, ip, nil
+		}
+		time.Sleep(retryInterval)
+	}
+
+	return nil, "", fmt.Errorf("timeout waiting for IP address")
+}
+
+func (l *LoadBalancerManager) checkPortOverlap(primary, secondary *v1.Service) error {
+	portMap := make(map[int32]bool)
+	for _, port := range secondary.Spec.Ports {
+		portMap[port.Port] = true
+	}
+
+	svcs, err := l.localSvcCache.List(metav1.NamespaceAll, labels.NewSelector())
```

Comment:
can we list all related svc by the primary svc label since only svcs with the same primary svc will share the same IP?

Reply:
Normally, we use annotations to configure parameters for the cloud provider.